src/core/modules/impl/CacheableContractDefinitionLoader.ts

```diff
@@ -0,0 +1,129 @@
+    this.srcCache = new LevelDbCache<SrcCache>({
+      ...cacheOptions,
+      dbLocation: `${cacheOptions.dbLocation}/source`
+    });
+  }
+
+  async load<State>(contractTxId: string, evolvedSrcTxId?: string): Promise<ContractDefinition<State>> {
+    const result = await this.getFromCache<State>(contractTxId, evolvedSrcTxId);
+    if (result) {
+      this.rLogger.debug('Hit from cache!', contractTxId, evolvedSrcTxId);
+      // LevelDB serializes Buffer to an object with 'type' and 'data' fields
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      if (result.contractType == 'wasm' && (result.srcBinary as any).data) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        result.srcBinary = Buffer.from((result.srcBinary as any).data);
+      }
+      this.verifyEnv(result);
+      return result;
+    }
+    const benchmark = Benchmark.measure();
+    const contract = await this.contractDefinitionLoader.load<State>(contractTxId, evolvedSrcTxId);
+
+    this.rLogger.info(`Contract definition loaded in: ${benchmark.elapsed()}`);
+    this.verifyEnv(contract);
+
+    await this.putToCache(contractTxId, contract, evolvedSrcTxId);
+
+    return contract;
+  }
+
+  async loadContractSource(contractSrcTxId: string): Promise<ContractSource> {
+    return await this.contractDefinitionLoader.loadContractSource(contractSrcTxId);
+  }
+
+  type(): GW_TYPE {
+    return this.contractDefinitionLoader.type();
+  }
+
+  setCache(cache: BasicSortKeyCache<ContractCache<unknown>>): void {
+    this.definitionCache = cache;
+  }
+
+  setSrcCache(cacheSrc: BasicSortKeyCache<SrcCache>): void {
+    this.srcCache = cacheSrc;
+  }
+
+  getCache(): BasicSortKeyCache<ContractCache<unknown>> {
+    return this.definitionCache;
+  }
+
+  getSrcCache(): BasicSortKeyCache<SrcCache> {
+    return this.srcCache;
+  }
+
+  private verifyEnv(def: ContractDefinition<unknown>): void {
+    if (def.testnet && this.env !== 'testnet') {
+      throw new Error('Trying to use testnet contract in a non-testnet env. Use the "forTestnet" factory method.');
+    }
+    if (!def.testnet && this.env === 'testnet') {
+      throw new Error('Trying to use non-testnet contract in a testnet env.');
+    }
+  }
+
+  // Gets ContractDefinition and ContractSource from two caches and returns a combined structure
+  private async getFromCache<State>(contractTxId: string, srcTxId?: string): Promise<ContractDefinition<State> | null> {
+    const contract = (await this.definitionCache.get(new CacheKey(contractTxId, 'cd'))) as SortKeyCacheResult<
+      ContractCache<State>
+    >;
+
+    if (!contract) {
+      return null;
+    }
+    const effectiveSrcTxId = srcTxId || contract.cachedValue.srcTxId;
+
+    const src = await this.srcCache.get(new CacheKey(effectiveSrcTxId, 'src'));
+    if (!src) {
+      return null;
+    }
+    return { ...contract.cachedValue, ...src.cachedValue, srcTxId: effectiveSrcTxId };
```

Comment:
Changes in this PR causes `WarpFactory.forLocal()` to use caching be default.

As a result almost all tests suddenly use caching... 
It turns out that this test failed `whitelist-sources-foreign-read.test.ts` becasue cache returned the wrong src tx id.
`srcTxId: effectiveSrcTxId` -> this is something that fixes this problem and maybe it should have been here all along?